casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala

```diff
@@ -0,0 +1,477 @@
+      runtimeManager: RuntimeManager,
+      validatorId: Option[ValidatorIdentity],
+      validators: Set[ByteString])(implicit scheduler: Scheduler)
+      extends CasperPacketHandlerInternal[F] {
+    private val nonePacket: F[Option[Packet]] = Applicative[F].pure(None: Option[Packet])
+
+    override def handleBlockMessage(bm: BlockMessage): F[Option[Packet]] = nonePacket
+    override def handleBlockRequest(peer: PeerNode, br: BlockRequest): F[Option[Packet]] =
+      nonePacket
+    override def handleApprovedBlockRequest(peer: PeerNode,
+                                            br: ApprovedBlockRequest): F[Option[Packet]] =
+      Option(noApprovedBlockAvailable(peer)).pure[F]
+
+    override def handleUnapprovedBlock(peer: PeerNode, ub: UnapprovedBlock): F[Option[Packet]] =
+      nonePacket
+    override def handleBlockApproval(ba: BlockApproval): F[Option[Packet]] = nonePacket
+
+    override def handleApprovedBlock(
+        ab: ApprovedBlock): F[(Option[MultiParentCasper[F]], Option[Packet])] =
+      for {
+        isValid <- Validate.approvedBlock[F](ab, validators)
+        casper <- if (isValid) {
+                   for {
+                     _            <- Log[F].info("CASPER: Valid ApprovedBlock received!")
+                     blockMessage = ab.candidate.flatMap(_.block).get
+                     _            <- BlockStore[F].put(blockMessage.blockHash, blockMessage)
+                     internalMap  <- BlockStore[F].asMap()
+                     _            <- LastApprovedBlock[F].set(ab)
+                     casper = MultiParentCasper
+                       .hashSetCasper[F](runtimeManager, validatorId, blockMessage, internalMap)
+                   } yield Option(casper)
+                 } else
+                   Log[F]
+                     .info("CASPER: Invalid ApprovedBlock received; refusing to add.")
+                     .map(_ => none[MultiParentCasper[F]])
+      } yield (casper, none[Packet])
+
+    override def handleNoApprovedBlockAvailable(na: NoApprovedBlockAvailable): F[Option[Packet]] =
+      for {
+        _ <- Log[F].info(s"No approved block available on node ${na.nodeIdentifer}")
+      } yield none[Packet]
+
+  }
+
+  /** Node in this state has already received at least one [[ApprovedBlock]] and it has created an instance
+    * of [[MultiParentCasper]].
+    *
+    * In the future it will be possible to create checkpoint with new [[ApprovedBlock]].
+    **/
+  class ApprovedBlockReceivedHandler[
+      F[_]: RPConfAsk: BlockStore: Monad: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler](
+      private val casper: MultiParentCasper[F],
+      approvedBlock: ApprovedBlock)
+      extends CasperPacketHandlerInternal[F] {
+
+    implicit val _casper = casper
+
+    private val nonePacket: F[Option[Packet]] = Applicative[F].pure(None: Option[Packet])
+
+    // Possible optimization in the future
```

Comment:
What is the possible optimization in the future?

Reply:
Having _checkpoints_ on the blockchain? Approving blocks different than genesis? Is that _legit_ ?